dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -0,0 +1,521 @@
+        // dimensions equally by the maximum amount that doesn't exceed the available size in either dimension.
+        // If the available size is infinite in both dimensions then we don't scale the visual. Otherwise, we
+        // calculate the scale factor by comparing the default visual size to the available size. This produces 2
+        // scale factors, one for each dimension. We choose the smaller of the scale factors to not exceed the
+        // available size in that dimension.
+        auto const visualSize = visual.Size();
+        if (visualSize != winrt::float2::zero())
+        {
+            const auto widthScale = availableSize.Width == std::numeric_limits<double>::infinity() ? std::numeric_limits<float>::infinity() : availableSize.Width / visualSize.x;
+            const auto heightScale = availableSize.Height == std::numeric_limits<double>::infinity() ? std::numeric_limits<float>::infinity() : availableSize.Height / visualSize.y;
+            if (widthScale == std::numeric_limits<double>::infinity() && heightScale == std::numeric_limits<double>::infinity())
+            {
+                return winrt::Size{ visualSize.x, visualSize.y };
+            }
+            else if (widthScale == std::numeric_limits<double>::infinity())
+            {
+                return winrt::Size{ visualSize.x * heightScale, availableSize.Height };
+            }
+            else if (heightScale == std::numeric_limits<double>::infinity())
+            {
+                return winrt::Size{ availableSize.Width, visualSize.y * widthScale };
+            }
+            else
+            {
+                return (heightScale > widthScale)
+                    ? winrt::Size{ availableSize.Width, visualSize.y * widthScale }
+                : winrt::Size{ visualSize.x * heightScale, availableSize.Height };
+            }
+        }
+        return visualSize;
+    }
+    // If we don't have a visual, we will show the fallback icon, so we need to do a traditional measure.
+    else
+    {
+        return __super::MeasureOverride(availableSize);
+    }
+}
+
+winrt::Size AnimatedIcon::ArrangeOverride(winrt::Size const& finalSize)
+{
+    if (auto const visual = m_animatedVisual.get())
+    {
+        auto const visualSize = visual.Size();
+        auto const scale = [finalSize, visual]()
+        {
+            auto scale = static_cast<winrt::float2>(finalSize) / visual.Size();
+            if (scale.x < scale.y)
+            {
+                scale.y = scale.x;
+            }
+            else
+            {
+                scale.x = scale.y;
+            }
+            return scale;
+        }();
+
+        winrt::float2 const arrangedSize = {
+            std::min(finalSize.Width / scale.x, visualSize.x),
+            std::min(finalSize.Height / scale.y, visualSize.y)
+        };
+        const auto offset = (finalSize - (visualSize * scale)) / 2;
```

Comment:
Is this the right thing to do when the alignments are HorizontalAlignment.Right and VerticalAlignement.Bottom for instance?   You always center it.

Reply:
I think so, this is the offset for the composition visual that is being inserted into the the animated icon.  HorizontalAlignment.Right on the animated icon will still arrange the control to the right of its layout window. But the visual will be centered within that.

Reply:
OK - I thought you may want to place the composition visual on the right/bottom too. You know better.